core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java

```diff
@@ -0,0 +1,234 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This is a {@link UfsAbsentPathCache} which implements asynchronous addition and removal to the
+ * cache, since the processing of the path may be slow.
+ */
+@ThreadSafe
+public final class AsyncUfsAbsentPathCache implements UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncUfsAbsentPathCache.class);
+  /** Number of threads for the async pool. */
+  private static final int NUM_THREADS = 50;
+  /** Number of seconds to keep threads alive. */
+  private static final int THREAD_KEEP_ALIVE_SECONDS = 60;
+  /** Number of paths to cache. */
+  private static final int MAX_PATHS =
+      Configuration.getInt(PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY);
+
+  /** The mount table. */
+  private final MountTable mMountTable;
+  /** Paths currently being processed. This is used to prevent duplicate processing. */
+  private final ConcurrentHashMapV8<String, ReadWriteLock> mCurrentPaths;
+  /** Cache of paths which are absent in the ufs. */
+  private final Cache<String, Long> mCache;
+  /** A thread pool for the async tasks. */
+  private final ThreadPoolExecutor mPool;
+
+  /**
+   * Creates a new instance of {@link AsyncUfsAbsentPathCache}.
+   *
+   * @param mountTable the mount table
+   */
+  public AsyncUfsAbsentPathCache(MountTable mountTable) {
+    mMountTable = mountTable;
+    mCurrentPaths = new ConcurrentHashMapV8<>(8, 0.95f, 8);
+    mCache = CacheBuilder.newBuilder().maximumSize(MAX_PATHS).build();
+
+    mPool = new ThreadPoolExecutor(NUM_THREADS, NUM_THREADS, THREAD_KEEP_ALIVE_SECONDS,
+        TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
+        ThreadFactoryUtils.build("UFS-Absent-Path-Cache-%d", true));
+  }
+
+  @Override
+  public void process(AlluxioURI path) {
+    mPool.submit(new ProcessPathTask(path));
+  }
+
+  @Override
+  public boolean isAbsent(AlluxioURI path) {
+    MountInfo mountInfo = getMountInfo(path);
+    if (mountInfo == null) {
+      return false;
+    }
+    AlluxioURI mountBaseUri = mountInfo.getAlluxioUri();
+
+    while (path != null && !path.equals(mountBaseUri)) {
+      Long cached = mCache.getIfPresent(path.getPath());
+      if (cached != null && cached == mountInfo.getMountId()) {
+        return true;
+      }
+      path = path.getParent();
+    }
+    // Reached the root, without finding anything in the cache.
+    return false;
+  }
+
+  /**
+   * Checks the existence of the corresponding ufs path for the given Alluxio path.
```

Comment:
this method has side-effects (e.g. cache invalidation or insertion), calling it `check` does not communicate this well ... I would call it `process`